Feature/9 solution review problem list

jd99iam opened the pull request:
## #9 

### 💡 PR 타입(하나 이상의 PR 타입을 선택해주세요)
* [x] 기능 추가
* [ ] 기능 삭제
* [ ] 버그 수정
* [ ] 의존성, 환경 변수, 빌드 관련 코드 업데이트

### 📲 반영 브랜치
feature/9-solution-review-problem-list -> 1-develop


### ❔ 변경 사항
풀이 리뷰 목록 기능 추가
문제 목록 기능 추가 

### 🔍 테스트 결과
- Repo 테스트
댓글리포지토리테스트, 문제리포지토리테스트, 풀이리포지토리테스트 정상작동
- Service 테스트
댓글서비스테스트, 풀이서비스테스트 정상작동

### 📰 참고 Reference
 
### ⚒️ 기타 사항
스프링 빈 순환 종속성 문제로 서버가 안돌아갑니다
![image](https://github.com/rlfrkdms1/Prog-rangers/assets/96513365/79acafff-2c34-4fd1-88b7-995415651a13)

jd99iam committed:
feat: 상세보기 한 줄 리뷰 서비스 추가

Related to : #9

jd99iam committed:
feat: ReviewRepository 추가

Related to : #9

jd99iam committed:
feat: ReviewRepository에 findAllBySolution 추가

Related to : #9

jd99iam committed:
feat: SolutionReviewsResponse 추가

Related to : #9

jd99iam committed:
feat: SolutionReviewsResponse에 필요한 dto들 추가

Related to : #9

jd99iam committed:
feat: SolutionService에 리뷰 반환 서비스 추가

1. SolutionService에 getReviewDetail 추가

Related to : #9

jd99iam committed:
feat: Member Entity에 Photo 추가

Related to : #9

jd99iam committed:
feat: SolutionService에 SolutionReviewResponse 반환하도록 추가

1. 반환형 수정

Related to : #9

jd99iam committed:
feat: SolutionController에서 SolutionReviewResponse 반환하도록 변경

Related to : #9

jd99iam committed:
refactor: SolutionController pathvariable 이름수정

1. soultionId -> solutionId 오타수정

Related to : #9

jd99iam committed:
feat: ProblemController 수정

1. problems가 쿼리스트링을 받도록 수정

Related to : #9

jd99iam committed:
feat: ProblemController 수정

1. problems가 쿼리스트링을 받도록 수정

Related to : #9

jd99iam committed:
feat: Problem 수정

1. problem에 date 필드 추가

Related to : #9

jd99iam committed:
feat: ProblemController에서 쿼리 파라미터 받지 않도록 수정

Related to : #9

jd99iam committed:
feat: ProblemRepository 수정

Related to : #9

jd99iam committed:
feat: ProblemService 수정

1. ProblemService에서 날짜 내림차순으로 받도록 수정

Related to : #9

jd99iam committed:
feat: ProblemService 페이지 네이션 추가

1. 페이지를 쿼리 스트링으로 받아서 해당 페이지의 문제들을 dto에 담아서 반환하도록 수정

Related to : #9

jd99iam committed:
feat: SolutionRequest에서 Problem 날짜 추가하도록 변경

Related to : #9

jd99iam committed:
fix: ProblemRepository에서 사용하는 메소드 명 변경

1.  findAllByOrderByDateDesc() 로 변경

Related to : #9

jd99iam committed:
fix: ProblemService 페이지 기능 추가

1.  쿼리스트링으로 받은 페이지에 해당하는 문제만 반환하도록 변경

Related to : #9

jd99iam committed:
fix: ProblemController, ProblemService 쿼리 파라미터 수정

1.  쿼리 파라미터로 페이지, 알고리즘, 자료구조, 정렬기준 받도록 수정

Related to : #9

jd99iam committed:
refactor: SolutionServiceTest 에서 빌더 사용하도로 변경

Related to : #9

jd99iam committed:
refactor: SolutionRepositoryTest 에서 빌더 사용하도로 변경

Related to : #9

jd99iam committed:
fix: 이미 존재하는 문제의 풀이일 경우 해당 풀이를 가져와 설정하도록 변경

Related to : #9

jd99iam committed:
feat: ProblemRepository 필터링에 필요한 쿼리 작성

Related to : #9

jd99iam committed:
fix: ProblemController의 requestParam required false로 수정

Related to : #9

jd99iam committed:
fix: ProblemController 쿼리스트링 받는 타입을 상수로 변경

Related to : #9

jd99iam committed:
feat: ProblemService 개발

1. Problem의 알고리즘, 자료구조 상위 3개를 골라오게 변경
2. 매핑 수정  필요

Related to : #9

jd99iam committed:
feat: ProblemService 매핑 수정

Related to : #9

jd99iam committed:
feat: ProblemService 매핑 수정

1. 풀이목록에서 상위 3개의 알고리즘과 상위 3개의 자료구조를 반환하도록 수정

Related to : #9

jd99iam committed:
refactor: DATA_STRUCTURE ENUM rename

enum type 이름 재정의

Related to : #1

jd99iam committed:
refactor: 패키지 구조 수정

repository 패키지 구조를 도메인별로 나뉘도록 수정했습니다

Related to : #9

jd99iam committed:
feat: querydsl 설정 추가

querydsl 설정을 추가했습니다

Related to : #9

jd99iam committed:
feat: ReviewRepository 리포지토리 구조 수정

ReviewRepository 리포지토리 구성을 수정했습니다

Related to : #9

jd99iam committed:
feat: ProblemRepository 리포지톨리 구조 수정

ProblemRepository 리포지토리 구성을 수정했습니다

Related to : #9

jd99iam committed:
feat: QueryDslProblemRepositoryImpl 자료구조, 알고리즘, 페이징 구현

자료구조, 알고리즘, 페이징에 의한 동적쿼리를 작성했습니다

Related to : #9

jd99iam committed:
feat: ProblemController, ProblemService 변경

ProblemController와 ProblemService에서 querydsl을 사용하도록 변경했습니다

Related to : #9

jd99iam committed:
feat: Querydsl 설정 완료

서버 돌아가도록 되도록 설정 완료

Related to : #9

jd99iam committed:
feat: ProblemRepository에서 fetchJoin 사용하도록 변경

그냥 join을 사용하자 entity를 찾을 수 없다는 오류가 발생해서 fetchJoin을 사용하도록 변경

Related to : #9

jd99iam committed:
feat: QueryDslProblemRepositoryImpl 쿼리문 수정

1. from(problem,solution)으로 넣어서 join(solution.problem)이 가능하도록 수정
2. alias가 겹치지 않도록 subproblem 사용

Related to : #9

jd99iam committed:
feat: QueryDslProblemRepositoryImpl 쿼리문 세타조인으로 수정

1. problem->solution 방향으로는 연관관계가 없기 때문에 세타조인으로 수정

Related to : #9

jd99iam committed:
feat: Problem -> Solution 연관관계 추가

1. 조회에 필요한 연관관계를 추가해서 양방향으로 수정

Related to : #9

jd99iam committed:
feat: ProblemResponse의 필드 수정

1. 조회에 필요한 필드를 Object 형으로 수정

Related to : #9

jd99iam committed:
feat: 문제에 태그 다는 기능 추가

1. 리포지토리에서 문제에 달 태그를 계산하는 코드 추가

Related to : #9

jd99iam committed:
fix: 문제에 태그다는 기능 수정

1. 태그가 null 일 경우 추가하지 않음
2. 내림차순으로 3개까지 추가하도록 변경

Related to : #9

jd99iam committed:
feat: problem 정렬 기준 추가

1. 문제 생성 기준 정렬 추가
2. solution 개수 생성 기준 추가

Related to : #9

jd99iam committed:
feat: problem 정렬 기준 추가

1. 문제 생성 기준 정렬 추가
2. solution 개수 생성 기준 추가

Related to : #9

jd99iam committed:
feat: LanguageConstant 추가

Solution을 필터링 할 때 언어가 필요해서 LanguageConstant 추가

Related to : #9

jd99iam committed:
feat: LanguageNotFoundException 추가

언어 타입이 존재하지 않을 때 터질 예외 추가

Related to : #9

jd99iam committed:
refactor: LanguageConstant 함수명 rename

getKrName을 getValue로 rename

Related to : #9

jd99iam committed:
feat: LEVEL_NOT_EXISTS 에러 추가

레벨이 존재하지 않을 때에 대한 에러코드 추가

Related to : #9

jd99iam committed:
feat: Solution에 language 필드 추가

풀이 언어 필터링을 위한 language 필드 추가

Related to : #9

jd99iam committed:
refactor: LanguageConstant 오타 수정

Related to : #9

jd99iam committed:
feat: SolutionController 에 solutionList 반환하는 메서드 추가

Related to : #9

jd99iam committed:
feat: SolutionListResponse 추가

Solution 목록보기 반환을 위한 dto SolutionListResponse 추가

Related to : #9

jd99iam committed:
feat: SolutionResponse 추가

Solution 목록보기 반환을 위한 dto SolutionResponse 추가

Related to : #9

jd99iam committed:
feat: SolutionController solutionList에 쿼리파라미터 추가

필터링, 정렬을 위한 쿼리파라미터 추가

Related to : #9

jd99iam committed:
feat: SolutionController solutionList에 쿼리파라미터 추가

필터링, 정렬을 위한 쿼리파라미터 추가

Related to : #9

jd99iam committed:
feat: SolutionListResponse 에 생성자 추가

Related to : #9

jd99iam committed:
feat: SolutionService getSolutionList 추가

풀이 목록보기 기능 서비스에 추가

Related to : #9

jd99iam committed:
refactor: SolutionService 에서 풀이 목록 반환에서 정적 팩토리 메소드 사용하도록 리팩터링

Related to : #9

jd99iam committed:
refactor: SolutionRepository가 QueryDslSolutionRepository 상속하도록 변경

1. QueryDslSolutionRepository 추가
2. QueryDslSolutionRepositoryImpl 추가

Related to : #9

jd99iam committed:
feat: QueryDslSolutionRepositoryImpl 에 getSolutionList 추가

언어, 알고리즘, 자료구조에 따른 필터링과 정렬기준에 따라 정렬해서 반환하는 메서드 추가

Related to : #9

jd99iam committed:
feat: 조건에 해당하는 Solution이 없는 경우 추가

조건에 해당하는 Solution이 없으면 null을 반환하도록 한다

Related to : #9

jd99iam committed:
feat: 문제에 대한 풀이를 찾는 것이므로 problemId 추가

Related to : #9

jd99iam committed:
feat: SolutionListResponse 메소드에서 problemId 받도록 변경

Related to : #9

jd99iam committed:
feat: ProblemId도 풀이 검색 조건에 반영하도록 변경

Related to : #9

jd99iam committed:
feat: SolutionController 풀이목록보기 매핑 변경

problemId에 대해 풀이목록 반환하는 것이기 때문에 API 명세에 맞게 requestmapping 변경

Related to : #9

jd99iam committed:
feat: ProblemRepository 쿼리에 페이징 추가

Related to : #9

jd99iam committed:
feat: Solution 목록보기 쿼리에 페이징 추가

Related to : #9

jd99iam committed:
feat: Solution 생성 시 언어 입력하도록 변경

Related to : #9

jd99iam committed:
refactor: Problem 목록조회 정적 팩토리 메소드 사용하도록 변경

Related to : #9

jd99iam committed:
refactor: Problem 목록조회 정적 팩토리 메소드 사용하도록 변경

Related to : #9

jd99iam committed:
refactor: ProblemRepository 사용하지 않는 메서드 삭제

Related to : #9

jd99iam committed:
refactor: ProblemRepository 사용하지 않는 메서드 삭제

Related to : #9

jd99iam committed:
chore: 패키지 구조 변경

SolutionResponse와 SolutionListResponse 위치 이동

Related to : #9

jd99iam committed:
refactor: Solution 생성 시 엔티티가 컨트롤러에서 노출되지 않도록 변경

Related to : #9

jd99iam committed:
feat: SolutionFixture 수정

고정된 필드는 유지하고 아닌 필드들은 메소드 파라미터로 받도록 수정

Related to : #9

jd99iam committed:
feat: ProblemFixture 추가

Related to : #9

jd99iam committed:
feat: MemberFixture 추가

Related to : #9

jd99iam committed:
feat: SolutionRequest에 toDto 메소드 추가

테스트 할 때 사용하기 위해 엔티티를 dto로 변환하는 메소드 추가

Related to : #9

jd99iam committed:
fix: ProblemFixture 문제 링크 검증 통과하도록 변경

Related to : #9

jd99iam committed:
refactor: 풀이 저장 테스트 픽스처 사용하도록 리팩토링

Related to : #9

jd99iam committed:
refactor: 스크랩 풀이 저장 테스트 픽스처 사용하도록 리팩토링

Related to : #9

jd99iam committed:
feat: 솔루션 픽스처 수정

1. 빌더 생성 메소드 추가
2. 빌더를 통해 스크랩 해서 솔루션 만드는 경우 추가
3. 빌더를 통해 일반 풀이 만드는 경우 추가

Related to : #9

jd99iam committed:
refactor: 수정 테스트 리팩터링

1. 픽스처를 사용하도록 리팩터링
2. when(any()).thenReturn().thenReturn().. 방식을 사용하도록 리팩터링

Related to : #9

jd99iam committed:
refactor: 풀이 서비스 풀이 삭제 테스트 리팩터링

1. 픽스처를 사용하도록 리팩터링
2. when(any()).thenReturn().thenReturn().. 방식을 사용하도록 리팩터링

Related to : #9

jd99iam committed:
feat: Comment 엔티티 수정

1. 불필요한 메서드 삭제
2. 수정하면 fixed가 true가 되도록 변경

Related to : #9

jd99iam committed:
feat: Comment 픽스처 생성

Related to : #9

jd99iam committed:
feat: SolutionFixture에 기본 솔루션 생성 메서드 추가

id만 받아서 기본 솔루션 생성하는 메서드 추가

Related to : #9

jd99iam committed:
feat: Comment에 updateSolution 메서드 추가

삭제하기 전에 연관관계 끊기 위해 updateSolution 메서드 추가

Related to : #9

jd99iam committed:
feat: CommentFixture에 기본 댓글 생성 메서드 추가

Related to : #9

jd99iam committed:
refactor : CommentServiceTest 픽스처 사용하도록 리팩터링

Related to : #9

jd99iam committed:
fix: 예외 생성자 파라미터 위치 수정

Related to : #9

jd99iam committed:
refactor: SolutionRepositoryTest 픽스처 사용하도록 리팩터링

Related to : #9

jd99iam committed:
feat: MemberFixture와 ProblemFixture에 id값을 지정해서 픽스처 생성하는 메서드 추가

Related to : #9

jd99iam committed:
feat: Solution 엔티티 cascade 옵션 수정

1. Problem은 Solution이 생성될 때 같이 생성되야 한다
2. Member는 Solution과 생명주기를 다르게 관리해야 한다

Related to : #9

jd99iam committed:
refactor: TestConfig 추가

테스트 상황에서 빈을 주입받기 위한 TestConfig 파일 추가

Related to : #9

jd99iam committed:
test: SolutionRepositoryTest 테스트 코드 수정

cascade 옵션 수정에 따른 테스트 코드 수정

Related to : #9

jd99iam committed:
test: CommentRepositoryTest 테스트 코드 픽스처 사용하도록 리팩터링

Related to : #9

jd99iam committed:
refactor: SolutionServiceTest 라인 정리

Related to : #9

jd99iam committed:
test: 문제 목록 조회 날짜순 정렬 테스트  추가

Related to : #9

jd99iam committed:
test: 문제 목록 조회 알고리즘, 자료구조 필터 테스트 추가

Related to : #9

jd99iam committed:
test: 문제 목록 조회 알고리즘, 자료구조 필터 테스트 추가

Related to : #9

jd99iam committed:
test: 문제 목록 조회 알고리즘, 자료구조 필터 테스트 추가

Related to : #9

jd99iam committed:
refactor: QueryDslSolutionRepositoryImpl 라인 정리

Related to : #9

jd99iam committed:
test: SolutionRepositoryTest 풀이 목록 조회 문제 id에 의한 조회 테스트 추가

Related to : #9

jd99iam committed:
test: SolutionRepositoryTest 풀이 목록 조회 알고리즘, 자료구조에 의한 조회 테스트 추가

Related to : #9

jd99iam committed:
test: SolutionFixture에 언어 포함해서 객체 생성 메서드 추가

Related to : #9

jd99iam committed:
test: SolutionRepositoryTest 언어로 필터링해서 조회 테스트 추가

Related to : #9

jd99iam committed:
test: SolutionRepositoryTest 풀이 목록 조회 페이징 테스트 추가

Related to : #9

jd99iam committed:
test: ProblemRepositoryTest 문제 목록 조회 페이징 테스트 추가

Related to : #9

jd99iam committed:
test: 테스트 코드 수정

1. 멤버 아이디를 null로 놓고 생성하는 픽스처로 수정
2. 문제 아이디를 생성된 문제에서 가져오는 방식으로 수정

Related to : #9

jd99iam committed:
fix: 충돌 해결, 빈 순환 참조 문제 발생

1. 브랜치 병합하면서 발생한 충돌 해결
2. 패키지 구조 수정
3. 빈 순환 참조 문제 발생, 아직 미해결

Related to : #9

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/controller/ProblemController.java:
```diff
@@ -23,9 +23,13 @@ public class ProblemController {
-        List<ProblemListResponse> problemListResponses = problemService.getProblemList();
-        return ResponseEntity.ok(problemListResponses);
+    public ResponseEntity<?> problems(
+            @RequestParam Integer page,
```
Pageable 안쓰고 그냥 `Integer`로 받은 이유가 있나요 ?

rlfrkdms1 replied:
처음 페이지를 호출할 땐 page 값이 없을 것같아요 ! required = false 로 하거나 default값을 설정하면 좋을 것 같아요

jd99iam replied:
> Pageable 안쓰고 그냥 `Integer`로 받은 이유가 있나요 ?

제가 Pageable에 대해 잘 몰라서 Integer를 사용했고 공부해보고 적용하겠습니다

jd99iam replied:
> 처음 페이지를 호출할 땐 page 값이 없을 것같아요 ! required = false 로 하거나 default값을 설정하면 좋을 것 같아요



> 처음 페이지를 호출할 땐 page 값이 없을 것같아요 ! required = false 로 하거나 default값을 설정하면 좋을 것 같아요

페이지 값이 없는 경우는 없을 것 같아서 default="1"로 주겠습니다. 의견 감사합니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/controller/SolutionController.java:
```diff
@@ -143,5 +148,24 @@ public ResponseEntity<?> deleteComment(@PathVariable Long solutionId, @PathVaria
 
+    // 상세보기 한 줄 리뷰
+    @GetMapping("{solutionId}/reviews")
+    public ResponseEntity<?> solutionReviews(@PathVariable Long solutionId){
```
solution의 review를 보는 건 ReviewController로 가는게 맞을 것 같은데 여기 있는 이유가 있나요 ?

jd99iam replied:
> solution의 review를 보는 건 ReviewController로 가는게 맞을 것 같은데 여기 있는 이유가 있나요 ?

review는 항상 solution에 속해있고, 'solution의' review를 보는 것이라 자연스럽게 이런 설계가 된 것 같은데 바꾸는게 좋을까요?

rlfrkdms1 replied:
review가 solution에 속해있더라도 review 라는 엔티티와 관련된 기능이라 변경하는게 좋을 것 같아보여요 ㅜㅜ

jd99iam replied:
> review가 solution에 속해있더라도 review 라는 엔티티와 관련된 기능이라 변경하는게 좋을 것 같아보여요 ㅜㅜ

ReviewController를 만들고 해당 컨트롤러로 메소드를 이동시켰습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/controller/SolutionController.java:
```diff
@@ -143,5 +148,24 @@ public ResponseEntity<?> deleteComment(@PathVariable Long solutionId, @PathVaria
+    // 상세보기 한 줄 리뷰
+    @GetMapping("{solutionId}/reviews")
+    public ResponseEntity<?> solutionReviews(@PathVariable Long solutionId){
+        SolutionReviewsResponse reviewDetail = solutionService.getReviewDetail(solutionId);
```
상세 리뷰를 가져오는 걸 solutionService에서 하는 이유가 있나요 ?

jd99iam replied:
> 상세 리뷰를 가져오는 걸 solutionService에서 하는 이유가 있나요 ?

리뷰를 가져오는 요청이 SolutionController에서 발생해서 이렇게 설계했는데, ReviewController로 바꾼다면 이 내용도 바꾸도록 하겠습니다

jd99iam replied:
> 상세 리뷰를 가져오는 걸 solutionService에서 하는 이유가 있나요 ?

수정했는데 수정하면서 발생하는 문제점이 하나 있습니다. 전 회의에서 SolutionNotFoundException 같이 id에 해당하는 걸 찾지 못했을 때 발생하는 예외를 Service의 메소드에서 던지기로 했었고, 또 서비스에서 다른 서비스를 의존하지 않기로 했던 걸로 기억합니다. 
`public class ReviewService {

    private final ReviewRepository reviewRepository;
    private final SolutionRepository solutionRepository;

    public SolutionReviewsResponse getReviewDetail(Long solutionId) {

        // solutionId에 해당하는 풀이 찾기
        Solution solution = solutionRepository.findById(solutionId).orElseThrow(SolutionNotFoundException::new);`

그러다 보니 ReviewService에서 Solution을 찾는 경우 SolutionService에 의존하거나(예외가 SolutionService에서 발생), SolutionRepository에 의존하거나 (ReviewService 에서 예외 발생) 두가지 중 하나여서 둘 다 지킬 수 없는 상황인 것 같은데 의견이 어떠신가요?

rlfrkdms1 replied:
reviewService에서 solution을 찾는 경우가 문제라는건가요 ?
이건 어쩔 수 없이 reviewService에서 solutionRepository로 찾아오고 예외도 던져야합니다.

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/dto/solution/SolutionDetailResponse.java:
SolutionDetailResponseComment만 response로 응답할 경우가 있나요 ?
그런 경우가 없고 SolutionDetailResponse안에서만 사용된다면 그냥 List<SolutionDetailComment>로 사용해도 될 것 같아요
만약 이게 싫고 Response가 붙고 싶다면 response라는 단어는 항상 끝에 붙으면 좋을 것 같아요 !
그리고 SolutionDetailResponseSolution이라는 작명 보다는 SolutionDetails라던지 다른 작명이 있을 것 같아요

jd99iam replied:
> SolutionDetailResponseComment만 response로 응답할 경우가 있나요 ? 그런 경우가 없고 SolutionDetailResponse안에서만 사용된다면 그냥 List로 사용해도 될 것 같아요 만약 이게 싫고 Response가 붙고 싶다면 response라는 단어는 항상 끝에 붙으면 좋을 것 같아요 ! 그리고 SolutionDetailResponseSolution이라는 작명 보다는 SolutionDetails라던지 다른 작명이 있을 것 같아요

그렇다면 말씀하신 대로 request, response라는 단어는 작명에서 제외하고 어디서 사용되는지 알기 쉽게dto.solution.response, dto.solution.request와 같이 dto 내에서 패키지를 나누면 어떨까요?

rlfrkdms1 replied:
request, response 라는 단어는 최종적으로 반환되는 DTO에만 붙어있으면 좋을 것 같아요 !!
SolutionDetailResponse는 그대로 두고, SolutionDetailiResponseComment가 SolutionDetailComment로 변경되는건 어떨까요 ?
패키지는 dto.request.solution은 어떨까요 ? dto.solution.request가 편한가요 ?

jd99iam replied:
> request, response 라는 단어는 최종적으로 반환되는 DTO에만 붙어있으면 좋을 것 같아요 !! SolutionDetailResponse는 그대로 두고, SolutionDetailiResponseComment가 SolutionDetailComment로 변경되는건 어떨까요 ? 패키지는 dto.request.solution은 어떨까요 ? dto.solution.request가 편한가요 ?

각 엔티티 별로 dto들이 모아져 있는게 개발할 때 보기 편할 것 같아서 저는 dto.solution.request 가 좋은 것 같습니다

jd99iam replied:
> request, response 라는 단어는 최종적으로 반환되는 DTO에만 붙어있으면 좋을 것 같아요 !! SolutionDetailResponse는 그대로 두고, SolutionDetailiResponseComment가 SolutionDetailComment로 변경되는건 어떨까요 ? 패키지는 dto.request.solution은 어떨까요 ? dto.solution.request가 편한가요 ?

![image](https://github.com/rlfrkdms1/Prog-rangers/assets/96513365/70d6f9ca-fb00-4cad-aecf-f9ec2dc74df5)

최종 요청, 응답에 사용되는 경우 뒤에 Request, Response를 붙였고, 
ex) SolutionDetailResponse
아닌 경우 뒤에 붙지 않도록 수정했습니다.
그리고 SolutionDetail에서 필요한 Solution이면 SolutionDetailSolution과 같이 수정했습니다.

추가 수정사항 필용하면 말씀해 주세요

rlfrkdms1 replied:
저희가 계층 구조라 request.solution을 제안한거기도 한데, 그게 편하시면 일단 그렇게 해주세요

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/dto/problem/ProblemAlgorithmDataStructureResponse.java:
어디서 사용되는 response인가요 ?

jd99iam replied:
> 어디서 사용되는 response인가요 ?

만들었다가 설계를 바꾸면서 사용하지 않게 된 dto라 삭제하겠습니다.

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/controller/SolutionController.java:
```diff
@@ -143,5 +148,24 @@ public ResponseEntity<?> deleteComment(@PathVariable Long solutionId, @PathVaria
+    }
 
+    // solution 목록보기
+    @GetMapping("{problemId}/solutions")
```
solution 목록 보기가 이렇게 되면 `/solutions/{problemId}/solutions`가 될 것 같은데 
`prog-rangers/problems/{problemId}/solutions가 돼야 할 것 같아요

jd99iam replied:
> solution 목록 보기가 이렇게 되면 `/solutions/{problemId}/solutions`가 될 것 같은데 `prog-rangers/problems/{problemId}/solutions가 돼야 할 것 같아요

문제의 풀이 목록을 보는 거니까 맞는 말씀인것 같습니다. ProblemController로 이동시키도록 하겠습니다. 감사합니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/controller/SolutionController.java:
이제 풀이는 problem을 통해 들어가니 RequestMapping("/solutions") 는 적절해 보이지 않은 것 같아요ㅜ

jd99iam replied:
> 이제 풀이는 problem을 통해 들어가니 RequestMapping("/solutions") 는 적절해 보이지 않은 것 같아요ㅜ

그럼 모든 풀이는 문제에 종속적이니까 SolutionController에서 RequestMapping("{problemId}/solutions") 형태로 묶는게 적절할까요?

rlfrkdms1 replied:
음 근데 만약 mypage의 solution 보기도 추가 된다면 Request Mapping은 필요 없어질 수도 있겠어요 ㅜ

jd99iam replied:
> 음 근데 만약 mypage의 solution 보기도 추가 된다면 Request Mapping은 필요 없어질 수도 있겠어요 ㅜ

일단 보류하겠습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/dto/problem/ProblemListResponse.java:
이전에 사용되던 DTO로 보이는데 어디에 사용되고 있나요 ?

jd99iam replied:
> 이전에 사용되던 DTO로 보이는데 어디에 사용되고 있나요 ?

삭제했습니다. 감사합니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/dto/problem/ProblemResponse.java:
```diff
@@ -0,0 +1,61 @@
+    List<Object> tags = new ArrayList<>();
+
+
+    public static List<ProblemResponse> toDto(List<Problem> results) {
```
toDto말고 of나 from과 같은 작명은 어떤가요

jd99iam replied:
일반적으로 of나 from을 사용하기 때문인가요??

rlfrkdms1 replied:
toDto라고 작명이 돼있는데 이게 사용될때 ProblemResponse.toDto(problems) 같은 형태로 사용될 것 같은데 부자연스럽지 않을까요 ? 엔티티로부터 DTO를 만드는 과정인데 DTO.toDTO(Entity)라는 작명은 조금 어색한 것 같다고 생각했어요 !!

jd99iam replied:
> toDto라고 작명이 돼있는데 이게 사용될때 ProblemResponse.toDto(problems) 같은 형태로 사용될 것 같은데 부자연스럽지 않을까요 ? 엔티티로부터 DTO를 만드는 과정인데 DTO.toDTO(Entity)라는 작명은 조금 어색한 것 같다고 생각했어요 !!

가은님 말씀 들어보니 from이 적절한 것 같아서 from으로 변경했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/dto/problem/ProblemResponse.java:
```diff
@@ -0,0 +1,61 @@
+                    .build();
+
+            List<Solution> solutions = result.getSolutions();
+            HashMap<Object,Integer> hm = new HashMap<>();
```
hm 말고 직관적인 작명으로 변경하면 좋을 것 같아요

rlfrkdms1 replied:
hm이라고 작명하니까 밑에 로직들이 어떤 로직인지 알기 어려워요

jd99iam replied:
> hm이라고 작명하니까 밑에 로직들이 어떤 로직인지 알기 어려워요

solutionAlgorithmCountMap으로 rename 했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/dto/problem/ProblemResponse.java:
```diff
@@ -0,0 +1,61 @@
+            List<Solution> solutions = result.getSolutions();
+            HashMap<Object,Integer> hm = new HashMap<>();
+
+            for (Solution solution : solutions) {
```
지금 solution을 모두 돌면서 상위 세개 태그를 가져오는 로직 같은데 이런 부분은 DTO 파일이 아니라 service나 repository에서 해결해야 할 로직같아 보여요
DTO 에선 간단한 변환 작업만 하는게 좋지 않을까요 ?

jd99iam replied:
> 지금 solution을 모두 돌면서 상위 세개 태그를 가져오는 로직 같은데 이런 부분은 DTO 파일이 아니라 service나 repository에서 해결해야 할 로직같아 보여요 DTO 에선 간단한 변환 작업만 하는게 좋지 않을까요 ?

service에서 repository에서의 결과를 토대로 문제의 풀이별로 상위 3개 태그들을 가져오고 dto에서는 변환만 하려면 데이터를 담고 전달하기 위해 적절한 dto가 또 필요하지 않을까요?

rlfrkdms1 replied:
음 솔직히 dto가 왜 필요한지는 잘 모르겠어요 ㅜ 
이 로직들을 그대로 service로 들고 가면 되는 것 아닌가요 ? ㅜ

jd99iam replied:
> 음 솔직히 dto가 왜 필요한지는 잘 모르겠어요 ㅜ 이 로직들을 그대로 service로 들고 가면 되는 것 아닌가요 ? ㅜ

말씀하신 대로 수정했습니다!

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/dto/solution/SolutionListResponse.java:
```diff
@@ -0,0 +1,46 @@
+            return null;
+        }
+        // 문제이름, 저지명 세팅
+        SolutionListResponse solutionListResponse = SolutionListResponse.builder()
```
세팅 과정이 왜 필요한건가요?

jd99iam replied:
해당 라인 아래 로직들은 같은 problem의 solution들에 대해 각각 SolutionResponse를 만드는 것이고 해당 라인은 그런 solution들이 공통으로 갖는 problem의 문제이름과 저지명을 가져와서 dto에 담는 과정입니다

rlfrkdms1 replied:
음 리스트의 첫번째 요소가 problem이고 그 다음부터는 solution들이라고 알기는 어려운 것 같아요 
다른 방식의 설계 방법은 없을까요 ?

jd99iam replied:
> 음 리스트의 첫번째 요소가 problem이고 그 다음부터는 solution들이라고 알기는 어려운 것 같아요 다른 방식의 설계 방법은 없을까요 ?

`        Problem problem = solutions.get(0).getProblem();

        // 문제이름, 저지명 세팅
        SolutionListResponse solutionListResponse = SolutionListResponse.builder()
                .problemName(problem.getTitle())
                .ojName(problem.getOjName())
                .solutionListSolutions(new ArrayList<>())
                .build();`

위와 같이 Problem을 변수로 지정해서 알기 쉽도록 변경했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/dto/solution/SolutionListResponse.java:
```diff
@@ -0,0 +1,46 @@
+                .solutions(new ArrayList<>())
+                .build();
+
+        for (Solution solution : solutions){
```
위에서 첫번째 solution을 세팅해줬는데 여기서도 첫번째 solution이 들어가지 않나요 ?

jd99iam replied:
> 위에서 첫번째 solution을 세팅해줬는데 여기서도 첫번째 solution이 들어가지 않나요 ?

이 반복문은 solution의 정보들을 가져와 각각 dto를 만드는 것이고 위의 세팅은 solution들의 문제(모두 같은 문제이므로) 0번째만 가져와서 문제의 이름과 저지명을 세팅한 것입니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/repository/problem/QueryDslProblemRepositoryImpl.java:
```diff
@@ -0,0 +1,63 @@
+                // solution을 조회해서 자료구조, 알고리즘을 알아내야 해서 성능을 위해 패치조인
+                .join(problem.solutions, solution).fetchJoin()
+                .where(dataStructureEq(dataStructure), algorithmEq(algorithm))
+                .orderBy(orderByWhat(orderBy))
```
orderByWhat말고 다른 작명은 없을까요 ? orderByWhat 의 파라미터가 명시적이면 괜찮은데 
orderBy(orderByWhat(orderBy))는 좀 이상해보여요ㅜㅜ

jd99iam replied:
> orderByWhat말고 다른 작명은 없을까요 ? orderByWhat 의 파라미터가 명시적이면 괜찮은데 orderBy(orderByWhat(orderBy))는 좀 이상해보여요ㅜㅜ

orderCondition으로 rename 했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/repository/solution/QueryDslSolutionRepositoryImpl.java:
```diff
@@ -0,0 +1,60 @@
+                .offset((page - 1) * 4)
+                .limit(4)
+                .fetch();
+        return result;
```
페이지네이션을 하려면 전체 페이지의 개수도 필요해요 !

jd99iam replied:
> 페이지네이션을 하려면 전체 페이지의 개수도 필요해요 !

회의 때 새로운 페이지를 넘길 때마다 새로운 요청을 줄 거라고 하셨는데 그러면 현재 페이지에 필요한 응답들만 있으면 되는 것 아닌가요?

rlfrkdms1 replied:
페이지 밑에 페이지를 나타내야해서 그래요 !!
1 | 2 | 3 | 4 | 5 이렇게요 !

jd99iam replied:
> 페이지 밑에 페이지를 나타내야해서 그래요 !! 1 | 2 | 3 | 4 | 5 이렇게요 !

그러면 전에 리포지토리에서 받아오는 dto를 따로 만들지 말자고 하셨는데, 쿼리 한방에 Solution이랑 개수까지 가져오려면 dto나 Tuple 형태로 가져와야 하는데, 기존 repository 쿼리는 놔두고 페이지 개수만 받아오는 리포지토리 메소드를 따로 만들어서 처리하는 게 맞을까요?

rlfrkdms1 replied:
Pageable 사용하면 해결 할 수 있습니다 !

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/ProblemService.java:
```diff
@@ -1,28 +1,32 @@
+            AlgorithmConstant algorithm,
+            DataStructureConstant dataStructure,
+            String sortBy) {
+        List<Problem> problems = problemRepository.searchByAlgorithmAndDataStructureOrderByDateDesc(page, dataStructure, algorithm, sortBy);
```
정렬이 꼭 date만으로 되는건 아니지 않나요 ?
problmeList를 조회할 때 항상 사용되는 것이니 repository의 메서드명을 그냥 findAll로 유지해도 될 것 같아요

jd99iam replied:
> 정렬이 꼭 date만으로 되는건 아니지 않나요 ? problmeList를 조회할 때 항상 사용되는 것이니 repository의 메서드명을 그냥 findAll로 유지해도 될 것 같아요

변경했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/SolutionService.java:
```diff
@@ -29,8 +43,18 @@ public class SolutionService {
-    public Long save(Solution solution) {
+    public Long save(SolutionRequest solutionRequest) {
+        Solution solution = solutionRequest.toEntity();
+        List<Problem> problems = problemRepository.findAllByLink(solution.getProblem().getLink());
```
링크로 조회되는 problem은 단건일텐데 findAll인 이유가 있나요 ?

jd99iam replied:
> 링크로 조회되는 problem은 단건일텐데 findAll인 이유가 있나요 ?

findByLink로 수정했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/SolutionService.java:
```diff
@@ -102,4 +126,82 @@ public SolutionDetailResponse getSolutionDetail(Long solutionId) {
+        Solution solution = findById(solutionId);
+        // 줄 나눠서 배열에 저장
+        String[] lines = solution.getCode().split("\n");
+        for (String line : lines) {
```
for문 안에 아무것도 없는데 필요한 구문인가요 ?

jd99iam replied:
> for문 안에 아무것도 없는데 필요한 구문인가요 ?

수정했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/SolutionService.java:
```diff
@@ -102,4 +126,82 @@ public SolutionDetailResponse getSolutionDetail(Long solutionId) {
+        for (String line : lines) {
+        }
+        // 최종 응답 dto에 풀이 내용을 넣는다
+        SolutionReviewsResponse solutionReviewsResponse = new SolutionReviewsResponse();
```
정적 팩토리 메서드 사용안하고 setter 사용한 이유가 있나요 ?

jd99iam replied:
> 정적 팩토리 메서드 사용안하고 setter 사용한 이유가 있나요 ?

정적 팩토리 메소드 from을 SolutionReviewResponse에 선언하고 활용하도록 변경했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/SolutionService.java:
```diff
@@ -102,4 +126,82 @@ public SolutionDetailResponse getSolutionDetail(Long solutionId) {
+        solutionReviewsResponse.setDataStructure(solution.getDataStructure());
+        // 먼저 최종 응답 dto에 각 라인을 넣는다
+        for (int i = 0; i < lines.length; i++) {
+            LineResponse lineResponse = LineResponse.builder()
```
이것도 메서드로 뺄 수 있을 것같아요

jd99iam replied:
> 이것도 메서드로 뺄 수 있을 것같아요

> 이것도 뺄 수 있을 것 같은데요 ? ReviewResponse.from(review)처럼요 !

```
        // 최종 응답 dto에 풀이 내용을 넣는다
        SolutionReviewsResponse solutionReviewsResponse = SolutionReviewsResponse.from(solution, lines);
```
와 같이 수정했습니다!

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/SolutionService.java:
```diff
@@ -102,4 +126,82 @@ public SolutionDetailResponse getSolutionDetail(Long solutionId) {
+                log.info("리뷰 id {}, 리뷰 내용 : {}, 부모 리뷰 id : {}", review.getId(), review.getContent(), review.getParentId());
+                // 부모가 없는 리뷰인 경우
+                if (review.getParentId() == null) {
+                    ReviewResponse reviewResponse = ReviewResponse.builder()
```
이것도 뺄 수 있을 것 같은데요 ? ReviewResponse.from(review)처럼요 !

jd99iam replied:
> 이것도 뺄 수 있을 것 같은데요 ? ReviewResponse.from(review)처럼요 !

ReviewResponse, ReplyResponse 모두 정적 팩토리 사용하도록 수정했습니다

rlfrkdms1 commented on backend/src/main/resources/application.properties:
application.properties는 올리지 않는게 어떨까요

jd99iam replied:
> application.properties는 올리지 않는게 어떨까요

.gitignore에 application.properties 올리지 않도록 추가했습니다

rlfrkdms1 commented on backend/src/test/java/com/prograngers/backend/repository/problem/ProblemRepositoryTest.java:
```diff
@@ -0,0 +1,201 @@
+        problemRepository.save(problem3);
+
+        // 풀이 : 문제 - 풀이를 join해서 가져오기 때문에 problem.s
+        Solution solution1 = 풀이1.일반_솔루션_생성(null, problem1, member, 0, BFS, QUEUE);
```
이거 다 합칠 수 있을 것 같은데요 ? 제가 보내드린 예시처럼 save 메서드를 알기 쉽게 따로 빼서 처리하면 좋을 것 같아요
Solution solution1 = 풀이 저장(풀이1.일반_솔루션_생성(null, problem1, member, 0, BFS, QUEUE));

jd99iam replied:
> 이거 다 합칠 수 있을 것 같은데요 ? 제가 보내드린 예시처럼 save 메서드를 알기 쉽게 따로 빼서 처리하면 좋을 것 같아요 Solution solution1 = 풀이 저장(풀이1.일반_솔루션_생성(null, problem1, member, 0, BFS, QUEUE));

말씀하신 대로 수정했습니다. 테스트코드가 훨씬 가독성이 좋아진 것 같아요

rlfrkdms1 reviewed (commented):
메서드들이 전체적으로 조금  긴 것 같아요 
적당한 메서드 추출이나, 메서드를 inline으로 넣는게 필요해보여요
고생많았습니다:)

jd99iam committed:
refactor: 라인 정리 및 불필요한 import 문 수정

Related to : #9

jd99iam committed:
feat: ProblemController의 problems 메서드에 page 값 default="1"로 설정

Related to : #9

jd99iam committed:
remove: 사용하지 않는 ProblemAlgorithmDataStructureResponse 삭제

Related to : #9

jd99iam committed:
refactor: Solution 목록보기 기능을 ProblemController로 이동

Related to : #9

jd99iam committed:
remove: ProblemListResponse.java 삭제

Related to : #9

jd99iam committed:
refactor: hm을 solutionAlgorithmCountMap으로 rename

Related to : #9

jd99iam committed:
refactor: orderByWhat을 orderCondition으로 rename

Related to : #9

jd99iam committed:
refactor: 문제 목록 조회 메소드 명 findAll로 변경

Related to : #9

jd99iam committed:
refactor: findAllByLink를 findByLink로 수정

link로 조회되는 Problem은 1개이기 때문에 findByLink로 수정

Related to : #9

jd99iam committed:
refactor: 불필요한 for문 삭제

Related to : #9

jd99iam committed:
refactor: SolutionReviewsResponse 생성 시 정적 팩토리 메소드 사용하도록 변경

Related to : #9

jd99iam committed:
refactor: SolutionReviewsResponse에 정적 팩토리 메소드 변경

Related to : #9

jd99iam committed:
docs: .gitigrnore에 application.properties 무시하도록 설정

Related to : #9

jd99iam committed:
refactor: ReplyResponse, ReviewResponse 정적 팩토리 메소드 사용하도록 수정

Related to : #9

jd99iam committed:
refactor: CommentRepositoryTest에 댓글 저장 테스트 댓글_저장 메소드 사용하도록 리팩터링

Related to : #9

jd99iam committed:
refactor: ProblemRepositoryTest에서 저장 메소드 사용하도록 리팩터링

Related to : #9

jd99iam committed:
refactor: SolutionRepositoryTest에서 저장 메소드 사용하도록 리팩터링

Related to : #9

jd99iam committed:
refactor: CommentServiceTest에서 저장 메소드 사용하도록 리팩터링

Related to : #9

jd99iam committed:
feat: 상세보기 한줄 리뷰를 ReviewController로 변경

Related to : #9

jd99iam committed:
refactor: SolutionController 라인 정리

Related to : #9

jd99iam committed:
feat: ReviewService 추가

1. 풀이의 리뷰 목록을 반환하기 위한 ReviewService 추가
2. SolutionService에서 해당 메서드 이동

Related to : #9

jd99iam committed:
feat: ReviewService 추가

1. 풀이의 리뷰 목록을 반환하기 위한 ReviewService 추가
2. SolutionService에서 해당 메서드 이동

Related to : #9

jd99iam committed:
refactor: dto package를 엔티티별로 request, response로 나눴다

Related to : #9

jd99iam committed:
refactor: dto 작명방식 변경

1. 최종 요청,응답인 경우 뒤에 Request, Response가 붙는다
2. 계층적으로 이름을 짓는다
예시) SolutionList 에서 필요한 Solution일경우 SolutionListSolution

Related to : #9

jd99iam committed:
refactor: 정적 팩토리 메서드 toDto에서 from으로 변경

Related to : #9

jd99iam committed:
refactor: dto에서는 dto로 변환만 하고 핵심 비즈니스 로직은 서비스에 있도록 수정

ProblemService로 알고리즘, 자료구조 3개를 찾는 로직을 옮겼다

Related to : #9

jd99iam committed:
refactor: createDto 메서드명 from으로 rename

Related to : #9

jd99iam committed:
refactor: Problem을 알 수 있도록 변수로 세팅

solutions.get(0)으로 지정하면 알기 어렵다는 의견 이 있어서 Problem을 변수로 뺐다

Related to : #9

jd99iam commented:
상세한 리뷰 감사합니다. 
Pageable은 추후에 구현 해보도록 하겠습니다
